tests/Controller/AdminProductVariantAjaxTest.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Tests\Controller;
+
+use ApiTestCase\JsonApiTestCase;
+use Symfony\Component\BrowserKit\Cookie;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken;
+
+final class AdminProductVariantAjaxTest extends JsonApiTestCase
+{
+    /** @test */
+    public function it_denies_access_to_product_variants_for_not_authenticated_user(): void
+    {
+        $this->client->request('GET', '/admin/ajax/product-variants/search-all');
+
+        $response = $this->client->getResponse();
+
+        $this->assertEquals($response->getStatusCode(), Response::HTTP_FOUND);
+    }
+
+    /** @test */
+    public function it_returns_only_specified_part_of_all_product_variants(): void
+    {
+        $this->loadFixturesFromFile('authentication/administrator.yml');
+        $this->loadFixturesFromFiles(['resources/product_variants.yml']);
+
+        $this->authenticateAdminUser();
+
+        $this->client->request('GET', '/admin/ajax/product-variants/search-all?phrase=');
```

Comment:
You didn't provide a phrase, is that intentional?

Reply:
Yes because real scenario is where phrase is parsed to empty string - I don't know why.

Reply:
I don't follow. What do you mean by "real scenario"?
I think there should be a test for both cases - with and without a phrase.

Reply:
In this case when you enter the edit page of specific catalog product and add a scope with type `for variants` and click on empty input - this is the moment when phrase become empty in SQL query even if its not empty as first.
I'll try to investigate that.
Tests are surely needed for it (maybe there are some already).

Reply:
Now I understand. In this case we're only missing the case when there's a search phrase.
I know that the original feature that did a search by phrase in a product context has a [Behat test](https://github.com/Sylius/Sylius/blob/master/features/product/managing_product_variants/autocomplete_product_variant_descriptor.feature). And there's a similar one for taxons.
Is an end-to-end with test with Behat necessary here or is it enough to have another integration test with PHPUnit?
@GSadee @Zales0123 @AdamKasp WDYT?

Reply:
IMHO another case in PHPUnit is a minimum, we have a case without specified `phrase` so the case with `phrase` also should be tested. And like You already mentioned we are testing autocompleting in different places so in Catalog Promotion also it should be tested. What is more, we even can make behat which will testing `limit` feature :)

Reply:
Limit is hardcoded and can be tested with PHPUnit too, but we would need more than 25 variants in fixtures.
To keep things simple I would go for another test with PHPUnit for now.
The other 2 tests I mentioned are API only, so there would no difference to the extent both tests cover the feature.
The main difference is in the format and I prefer PHPUnit.

I'm also wondering wherever AJAX endpoints like this one should be ported to the new api.

Reply:
👍 for another test in PHPUnit, Behat is a pain for auto-complete as they need to be `@javascript` scenarios 😄 Covering it nicely with PHPUnit should be enough 👍